File: src/main/java/com/spring/reproducer/spring/test/sleuth/reproducer/TracingResource.java
package com.spring.reproducer.spring.test.sleuth.reproducer;

import org.springframework.cloud.sleuth.CurrentTraceContext;
import org.springframework.cloud.sleuth.TraceContext;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;
import reactor.util.context.ContextView;


@RestController
@RequestMapping("/api/tracing")
public class TracingResource {

    private static final Class<TraceContext> KEY = TraceContext.class;
    private static final Class<CurrentTraceContext> CURRENT_KEY = CurrentTraceContext.class;

    @GetMapping("spanId")
    public Mono<String> spanId() {
        return Mono.deferContextual(view -> traceContext(view))
                .map( c -> c.spanId() );

    }

    @GetMapping("parentId")
    public Mono<String> parentId() {
        return Mono.deferContextual(view -> traceContext(view))
            .map(c -> c.parentId());
    }

    private Mono<TraceContext> traceContext(ContextView contextView){
        return Mono.justOrEmpty(contextView.get(KEY));
    }

    @GetMapping("currentSpanId")
    public Mono<String> spanIdForCurrentTraceContext() {
        return Mono.deferContextual(view -> currentTraceContext(view))
            .map(c -> c.context().spanId());

    }

    @GetMapping("currentParentId")
    public Mono<String> parentIdForCurrentTraceContext() {
        return Mono.deferContextual(view -> currentTraceContext(view))
            .map(c -> c.context().parentId());

    }

    private Mono<CurrentTraceContext> currentTraceContext(ContextView contextView){
        return Mono.justOrEmpty(contextView.get(CURRENT_KEY));
    }
}

File: src/main/kotlin/com/spring/reproducer/spring/test/sleuth/reproducer/TracingResource.kt
package com.spring.reproducer.spring.test.sleuth.reproducer

import org.springframework.cloud.sleuth.CurrentTraceContext
import org.springframework.cloud.sleuth.TraceContext
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import reactor.util.context.ContextView

@RestController
@RequestMapping("/api/tracing")
class TracingResource {

    companion object {
        val KEY = TraceContext::class.java
        val CURRENT_KEY = CurrentTraceContext::class.java
    }

    @GetMapping("spanId")
    fun spanId(): Mono<String> {
        return Mono.deferContextual { view -> view.traceContext() }
                .map { it.spanId() }

    }

    @GetMapping("parentId")
    fun parentId(): Mono<String> {
        return Mono.deferContextual { view -> view.traceContext() }
            .map { it.parentId() }

    }

    private fun ContextView.traceContext() = get(KEY).toMono()

    @GetMapping("currentSpanId")
    fun spanIdForCurrentTraceContext(): Mono<String> {
        return Mono.deferContextual { view -> view.currentTraceContext() }
            .map { it.context()?.spanId() }

    }

    @GetMapping("currentParentId")
    fun parentIdForCurrentTraceContext(): Mono<String> {
        return Mono.deferContextual { view -> view.currentTraceContext() }
            .map { it.context()?.parentId() }

    }

    private fun ContextView.currentTraceContext() = get(CURRENT_KEY).toMono()
}